src/relations/postgresql_provider.py

```diff
@@ -71,6 +84,27 @@ def _sanitize_extra_roles(extra_roles: str | None) -> list[str]:
         extra_roles_list = [role for role in extra_roles_list if role not in ACCESS_GROUPS]
         return extra_roles_list
 
+    def get_username_mapping(self) -> dict[str, str]:
+        """Get a mapping of custom usernames by a relation ID."""
+        if username_mapping := self.charm.get_secret(APP_SCOPE, USERNAME_MAPPING_LABEL):
+            return json.loads(username_mapping)
+        return {}
+
+    def update_username_mapping(self, relation_id: int, username: str | None) -> None:
```

Comment:
Add or delete a mapping between rel ID and custom username. Default generated usernames are ignored, since we can infer them based on rel ID as before.